File: services/elitea-worker-python/src/elitea_worker/security.py
"""File-backed production trust material and exact signing-key resolution."""

from __future__ import annotations

import base64
import binascii
import json
import ssl
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from pydantic import BaseModel, ConfigDict, Field, model_validator

from elitea_worker.config import RuntimeDeployConfig, read_regular_file
from elitea_worker.execution.errors import InvalidInput


_MAX_REDIS_PASSWORD_BYTES = 512
_MAX_REDIS_PASSWORD_FILE_BYTES = _MAX_REDIS_PASSWORD_BYTES + 2


class _PublicKeyEntry(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    key_id: str = Field(min_length=1, max_length=256)
    public_key_base64: str = Field(min_length=1, max_length=128)


class _PublicKeyring(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["elitea.runtime-ed25519-keyring.v1"]
    keys: tuple[_PublicKeyEntry, ...] = Field(min_length=1, max_length=64)

    @model_validator(mode="after")
    def validate_unique_safe_ids(self) -> _PublicKeyring:
        key_ids = [item.key_id for item in self.keys]
        if len(set(key_ids)) != len(key_ids):
            raise ValueError("duplicate signing key ID")
        if any(
            any(character in key_id for character in ("\r", "\n", "\x00"))
            for key_id in key_ids
        ):
            raise ValueError("signing key ID is malformed")
        return self


class ExactEd25519PublicKeyResolver:
    """Immutable keyring with no default, prefix or fallback lookup."""

    def __init__(self, keys: dict[str, Ed25519PublicKey]) -> None:
        if not keys or any(not key_id for key_id in keys):
            raise ValueError("at least one exact Ed25519 key is required")
        self._keys = dict(keys)

    def resolve_ed25519_public_key(self, key_id: str) -> Ed25519PublicKey:
        return self._keys[key_id]


@dataclass(frozen=True, slots=True)
class RuntimeTrustMaterial:
    ca_bytes: bytes
    certificate_bytes: bytes
    private_key_bytes: bytes
    spool_master_key: bytes
    redis_password: str
    signing_keys: ExactEd25519PublicKeyResolver
    ca_path: Path
    certificate_path: Path
    private_key_path: Path

    @classmethod
    def load(cls, config: RuntimeDeployConfig) -> RuntimeTrustMaterial:
        ca = read_regular_file(
            config.ca_path,
            max_bytes=1024 * 1024,
            private=False,
            description="runtime CA bundle",
        )
        certificate = read_regular_file(
            config.certificate_path,
            max_bytes=256 * 1024,
            private=False,
            description="workload certificate chain",
        )
        private_key = read_regular_file(
            config.private_key_path,
            max_bytes=128 * 1024,
            private=True,
            description="workload private key",
        )
        spool_key = read_regular_file(
            config.spool_key_path,
            max_bytes=32,
            private=True,
            description="output spool key",
        )
        if len(spool_key) != 32:
            raise InvalidInput("The output spool key is unavailable or unsafe.")
        redis_password_text = _load_redis_password(config.redis_password_path)
        resolver = load_ed25519_keyring(config.ed25519_keyring_path)
        # Parsing the chain now proves the configured certificate and key match.
        context = _base_client_context(config.ca_path)
        try:
            context.load_cert_chain(
                certfile=str(config.certificate_path),
                keyfile=str(config.private_key_path),
            )
        except (OSError, ssl.SSLError) as exc:
            raise InvalidInput("The workload TLS identity is invalid.") from exc
        return cls(
            ca_bytes=ca,
            certificate_bytes=certificate,
            private_key_bytes=private_key,
            spool_master_key=spool_key,
            redis_password=redis_password_text,
            signing_keys=resolver,
            ca_path=config.ca_path,
            certificate_path=config.certificate_path,
            private_key_path=config.private_key_path,
        )

    def http_client_context(self) -> ssl.SSLContext:
        context = _base_client_context(self.ca_path)
        try:
            context.load_cert_chain(
                certfile=str(self.certificate_path),
                keyfile=str(self.private_key_path),
            )
        except (OSError, ssl.SSLError) as exc:
            raise InvalidInput("The workload TLS identity is invalid.") from exc
        return context


def _base_client_context(ca_path: Path) -> ssl.SSLContext:
    try:
        # Do not merge machine/container system roots into this private-plane
        # trust domain. Only the explicitly deployed CA may authenticate it.
        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
        context.load_verify_locations(cafile=str(ca_path))
    except (OSError, ssl.SSLError) as exc:
        raise InvalidInput("The runtime CA bundle is invalid.") from exc
    context.minimum_version = ssl.TLSVersion.TLSv1_3
    context.check_hostname = True
    context.verify_mode = ssl.CERT_REQUIRED
    return context


def _load_redis_password(path: Path) -> str:
    raw = read_regular_file(
        path,
        max_bytes=_MAX_REDIS_PASSWORD_FILE_BYTES,
        private=True,
        description="Redis ACL password",
    )
    if raw.endswith(b"\n"):
        raw = raw[:-1]
        if raw.endswith(b"\r"):
            raw = raw[:-1]
    if (
        not 1 <= len(raw) <= _MAX_REDIS_PASSWORD_BYTES
        or any(marker in raw for marker in (b"\r", b"\n", b"\x00"))
    ):
        raise InvalidInput("The Redis ACL password is unavailable or unsafe.")
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise InvalidInput("The Redis ACL password is unavailable or unsafe.") from exc


def _parse_keyring(raw: bytes) -> ExactEd25519PublicKeyResolver:
    try:
        value = json.loads(raw)
        keyring = _PublicKeyring.model_validate(value)
        keys: dict[str, Ed25519PublicKey] = {}
        for item in keyring.keys:
            encoded = item.public_key_base64
            if any(character.isspace() for character in encoded):
                raise ValueError("public key encoding contains whitespace")
            key_bytes = base64.b64decode(encoded, validate=True)
            if len(key_bytes) != 32:
                raise ValueError("Ed25519 public key must contain 32 bytes")
            keys[item.key_id] = Ed25519PublicKey.from_public_bytes(key_bytes)
        return ExactEd25519PublicKeyResolver(keys)
    except (ValueError, TypeError, json.JSONDecodeError, binascii.Error) as exc:
        raise InvalidInput("The Ed25519 verification keyring is invalid.") from exc


def load_ed25519_keyring(path: Path) -> ExactEd25519PublicKeyResolver:
    raw = read_regular_file(
        path,
        max_bytes=64 * 1024,
        private=False,
        description="Ed25519 verification keyring",
    )
    return _parse_keyring(raw)

File: services/elitea-worker-rust/src/security.rs
//! File-backed production trust material and exact command-key resolution.

use std::collections::BTreeMap;
use std::fmt;
use std::sync::Arc;

use base64::Engine as _;
use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
use rustls::pki_types::{CertificateDer, PrivateKeyDer, pem::PemObject as _};
use rustls::{ClientConfig, RootCertStore};
use serde::Deserialize;
use tonic::transport::{Certificate, Identity};
use zeroize::Zeroizing;

use crate::config::{RuntimeConfigError, RuntimeDeployConfig, read_regular_file};
use crate::protocol::command::{
    Ed25519CommandAuthenticator, Ed25519PublicKeyResolver, SignedCommandAuthenticator,
};
use crate::spool::SpoolMasterKey;
use crate::transport::redis_streams::{RedisStreamsError, RedisTlsMaterial};

const KEYRING_SCHEMA_VERSION: &str = "elitea.runtime-ed25519-keyring.v1";
const MAX_CA_BYTES: usize = 1024 * 1024;
const MAX_CERTIFICATE_BYTES: usize = 256 * 1024;
const MAX_PRIVATE_KEY_BYTES: usize = 128 * 1024;
const MAX_KEYRING_BYTES: usize = 64 * 1024;
const MAX_REDIS_PASSWORD_FILE_BYTES: usize = 514;
const MAX_REDIS_PASSWORD_BYTES: usize = 512;
const MAX_KEY_ID_BYTES: usize = 256;
const MAX_KEYS: usize = 64;

/// Stable, redacted failure at the process trust boundary.
#[derive(Debug)]
pub enum RuntimeTrustError {
    Material(RuntimeConfigError),
    InvalidTlsIdentity,
    InvalidSpoolKey,
    InvalidRedisPassword,
    InvalidSigningKeyring,
    InvalidRedisTls,
}

impl fmt::Display for RuntimeTrustError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Material(_) => "runtime trust material is unavailable or unsafe",
            Self::InvalidTlsIdentity => "the workload TLS identity is invalid",
            Self::InvalidSpoolKey => "the output spool key is invalid",
            Self::InvalidRedisPassword => "the Redis ACL password is invalid",
            Self::InvalidSigningKeyring => "the Ed25519 verification keyring is invalid",
            Self::InvalidRedisTls => "the Redis TLS material is invalid",
        })
    }
}

impl std::error::Error for RuntimeTrustError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::Material(error) => Some(error),
            Self::InvalidTlsIdentity
            | Self::InvalidSpoolKey
            | Self::InvalidRedisPassword
            | Self::InvalidSigningKeyring
            | Self::InvalidRedisTls => None,
        }
    }
}

impl From<RuntimeConfigError> for RuntimeTrustError {
    fn from(value: RuntimeConfigError) -> Self {
        Self::Material(value)
    }
}

/// Exact immutable key-ID resolver with no default or prefix lookup.
#[derive(Clone)]
pub struct ExactEd25519PublicKeyResolver {
    keys: Arc<BTreeMap<Box<str>, [u8; 32]>>,
}

impl Ed25519PublicKeyResolver for ExactEd25519PublicKeyResolver {
    fn resolve_ed25519_public_key(&self, key_id: &str) -> Option<[u8; 32]> {
        self.keys.get(key_id).copied()
    }
}

/// Process-owned private-plane material for gRPC/HTTP, command verification
/// and encrypted output. Redis credentials are intentionally reloaded per
/// connection generation through [`load_redis_tls_material`].
pub struct RuntimeTrustMaterial {
    ca_pem: Vec<u8>,
    certificate_pem: Vec<u8>,
    private_key_pem: Zeroizing<Vec<u8>>,
    spool_key: Zeroizing<[u8; 32]>,
    signing_keys: ExactEd25519PublicKeyResolver,
}

impl RuntimeTrustMaterial {
    /// Load and validate the process-owned workload identity, output-spool key
    /// and exact command-signing keyring.
    ///
    /// # Errors
    ///
    /// Returns a redacted trust error when any file, permission, format,
    /// certificate/key relationship or keyring invariant is invalid.
    pub fn load(config: &RuntimeDeployConfig) -> Result<Self, RuntimeTrustError> {
        let ca_pem = read_regular_file(&config.ca_path, MAX_CA_BYTES, false, "runtime CA bundle")?;
        let certificate_pem = read_regular_file(
            &config.certificate_path,
            MAX_CERTIFICATE_BYTES,
            false,
            "workload certificate chain",
        )?;
        let private_key_pem = Zeroizing::new(read_regular_file(
            &config.private_key_path,
            MAX_PRIVATE_KEY_BYTES,
            true,
            "workload private key",
        )?);
        validate_tls_identity(&ca_pem, &certificate_pem, &private_key_pem)?;
        let spool_key = read_regular_file(&config.spool_key_path, 32, true, "output spool key")?;
        let spool_key: [u8; 32] = spool_key
            .try_into()
            .map_err(|_| RuntimeTrustError::InvalidSpoolKey)?;
        let signing_keys = load_ed25519_keyring(config)?;
        Ok(Self {
            ca_pem,
            certificate_pem,
            private_key_pem,
            spool_key: Zeroizing::new(spool_key),
            signing_keys,
        })
    }

    #[must_use]
    pub fn private_ca(&self) -> Certificate {
        Certificate::from_pem(self.ca_pem.clone())
    }

    #[must_use]
    pub fn client_identity(&self) -> Identity {
        Identity::from_pem(self.certificate_pem.clone(), &self.private_key_pem)
    }

    #[must_use]
    pub fn spool_master_key(&self) -> SpoolMasterKey {
        SpoolMasterKey::new(*self.spool_key)
    }

    #[must_use]
    pub fn command_authenticator(&self) -> Arc<dyn SignedCommandAuthenticator> {
        Arc::new(Ed25519CommandAuthenticator::new(self.signing_keys.clone()))
    }
}

/// Reload the Redis password and workload TLS files for one fresh connection
/// generation. The returned value owns zeroizing secret buffers.
///
/// # Errors
///
/// Returns a redacted trust error when any credential file or Redis-specific
/// password/TLS invariant is invalid.
pub fn load_redis_tls_material(
    config: &RuntimeDeployConfig,
) -> Result<RedisTlsMaterial, RuntimeTrustError> {
    let password = load_redis_password(config)?;
    let ca_pem = read_regular_file(&config.ca_path, MAX_CA_BYTES, false, "runtime CA bundle")?;
    let certificate_pem = read_regular_file(
        &config.certificate_path,
        MAX_CERTIFICATE_BYTES,
        false,
        "workload certificate chain",
    )?;
    let private_key_pem = read_regular_file(
        &config.private_key_path,
        MAX_PRIVATE_KEY_BYTES,
        true,
        "workload private key",
    )?;
    RedisTlsMaterial::new(password, ca_pem, certificate_pem, private_key_pem)
        .map_err(map_redis_tls_error)
}

fn load_redis_password(config: &RuntimeDeployConfig) -> Result<String, RuntimeTrustError> {
    let mut bytes = Zeroizing::new(read_regular_file(
        &config.redis_password_path,
        MAX_REDIS_PASSWORD_FILE_BYTES,
        true,
        "Redis ACL password",
    )?);
    if bytes.ends_with(b"\n") {
        bytes.pop();
        if bytes.ends_with(b"\r") {
            bytes.pop();
        }
    }
    if bytes.is_empty()
        || bytes.len() > MAX_REDIS_PASSWORD_BYTES
        || bytes
            .iter()
            .any(|byte| matches!(byte, b'\r' | b'\n' | b'\0'))
    {
        return Err(RuntimeTrustError::InvalidRedisPassword);
    }
    String::from_utf8(bytes.to_vec()).map_err(|_| RuntimeTrustError::InvalidRedisPassword)
}

fn load_ed25519_keyring(
    config: &RuntimeDeployConfig,
) -> Result<ExactEd25519PublicKeyResolver, RuntimeTrustError> {
    let raw = read_regular_file(
        &config.ed25519_keyring_path,
        MAX_KEYRING_BYTES,
        false,
        "Ed25519 verification keyring",
    )?;
    parse_ed25519_keyring(&raw)
}

fn parse_ed25519_keyring(raw: &[u8]) -> Result<ExactEd25519PublicKeyResolver, RuntimeTrustError> {
    let keyring = serde_json::from_slice::<PublicKeyring>(raw)
        .map_err(|_| RuntimeTrustError::InvalidSigningKeyring)?;
    if keyring.schema_version != KEYRING_SCHEMA_VERSION
        || keyring.keys.is_empty()
        || keyring.keys.len() > MAX_KEYS
    {
        return Err(RuntimeTrustError::InvalidSigningKeyring);
    }
    let mut keys = BTreeMap::new();
    for entry in keyring.keys {
        if entry.key_id.is_empty()
            || entry.key_id.len() > MAX_KEY_ID_BYTES
            || entry
                .key_id
                .bytes()
                .any(|byte| matches!(byte, b'\r' | b'\n' | b'\0'))
            || entry
                .public_key_base64
                .bytes()
                .any(|byte| byte.is_ascii_whitespace())
        {
            return Err(RuntimeTrustError::InvalidSigningKeyring);
        }
        let decoded = BASE64_STANDARD
            .decode(entry.public_key_base64)
            .map_err(|_| RuntimeTrustError::InvalidSigningKeyring)?;
        let public_key: [u8; 32] = decoded
            .try_into()
            .map_err(|_| RuntimeTrustError::InvalidSigningKeyring)?;
        if keys.insert(entry.key_id.into(), public_key).is_some() {
            return Err(RuntimeTrustError::InvalidSigningKeyring);
        }
    }
    Ok(ExactEd25519PublicKeyResolver {
        keys: Arc::new(keys),
    })
}

fn validate_tls_identity(
    ca_pem: &[u8],
    certificate_pem: &[u8],
    private_key_pem: &[u8],
) -> Result<(), RuntimeTrustError> {
    let mut roots = RootCertStore::empty();
    let ca_certificates = CertificateDer::pem_slice_iter(ca_pem)
        .collect::<Result<Vec<_>, _>>()
        .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?;
    if ca_certificates.is_empty() {
        return Err(RuntimeTrustError::InvalidTlsIdentity);
    }
    for certificate in ca_certificates {
        roots
            .add(certificate)
            .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?;
    }
    let client_certificates = CertificateDer::pem_slice_iter(certificate_pem)
        .collect::<Result<Vec<_>, _>>()
        .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?;
    if client_certificates.is_empty() {
        return Err(RuntimeTrustError::InvalidTlsIdentity);
    }
    let private_key = PrivateKeyDer::from_pem_slice(private_key_pem)
        .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?;
    ClientConfig::builder_with_provider(rustls::crypto::ring::default_provider().into())
        .with_protocol_versions(&[&rustls::version::TLS13])
        .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?
        .with_root_certificates(roots)
        .with_client_auth_cert(client_certificates, private_key)
        .map_err(|_| RuntimeTrustError::InvalidTlsIdentity)?;
    Ok(())
}

fn map_redis_tls_error(_error: RedisStreamsError) -> RuntimeTrustError {
    RuntimeTrustError::InvalidRedisTls
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct PublicKeyring {
    schema_version: String,
    keys: Vec<PublicKeyEntry>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct PublicKeyEntry {
    key_id: String,
    public_key_base64: String,
}

#[cfg(test)]
mod tests {
    use base64::Engine as _;
    use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
    use serde_json::json;

    use super::{RuntimeTrustError, parse_ed25519_keyring};
    use crate::protocol::command::Ed25519PublicKeyResolver;

    #[test]
    fn exact_keyring_has_no_default_prefix_or_duplicate_resolution() {
        let public_key = [7_u8; 32];
        let encoded = BASE64_STANDARD.encode(public_key);
        let raw = serde_json::to_vec(&json!({
            "schema_version": "elitea.runtime-ed25519-keyring.v1",
            "keys": [{"key_id": "runtime-signing-1", "public_key_base64": encoded}]
        }))
        .expect("keyring JSON");

        let resolver = parse_ed25519_keyring(&raw).expect("valid keyring");

        assert_eq!(
            resolver.resolve_ed25519_public_key("runtime-signing-1"),
            Some(public_key)
        );
        assert_eq!(resolver.resolve_ed25519_public_key("runtime-signing"), None);
        assert_eq!(resolver.resolve_ed25519_public_key("unknown"), None);
    }

    #[test]
    fn malformed_or_ambiguous_keyrings_fail_closed() {
        let encoded = BASE64_STANDARD.encode([9_u8; 32]);
        let cases = [
            json!({"schema_version": "wrong", "keys": [{"key_id": "one", "public_key_base64": encoded}]}),
            json!({"schema_version": "elitea.runtime-ed25519-keyring.v1", "keys": []}),
            json!({"schema_version": "elitea.runtime-ed25519-keyring.v1", "keys": [
                {"key_id": "one", "public_key_base64": encoded},
                {"key_id": "one", "public_key_base64": encoded}
            ]}),
            json!({"schema_version": "elitea.runtime-ed25519-keyring.v1", "keys": [
                {"key_id": "one\nleak", "public_key_base64": encoded}
            ]}),
            json!({"schema_version": "elitea.runtime-ed25519-keyring.v1", "keys": [
                {"key_id": "one", "public_key_base64": "AAAA\nAAAA"}
            ]}),
            json!({"schema_version": "elitea.runtime-ed25519-keyring.v1", "keys": [
                {"key_id": "one", "public_key_base64": BASE64_STANDARD.encode([1_u8; 31])}
            ]}),
        ];
        for value in cases {
            let raw = serde_json::to_vec(&value).expect("keyring JSON");
            assert!(matches!(
                parse_ed25519_keyring(&raw),
                Err(RuntimeTrustError::InvalidSigningKeyring)
            ));
        }
    }
}
